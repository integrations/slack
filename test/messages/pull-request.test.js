const { PullRequest } = require('../../lib/messages/pull-request');
const combinedStatus = require('../fixtures/combined_status.json');
const combinedStatusAllPassing = require('../fixtures/combined_status_all_passing.json');
const combinedStatusSomePassing = require('../fixtures/combined_status_some_passing.json');
const combinedStatusOneFailing = require('../fixtures/combined_status_one_failing.json');

const pullRequestOpened = require('../fixtures/webhooks/pull_request.opened.json');
const pullRequestClosed = require('../fixtures/webhooks/pull_request.closed.json');

describe('Pull request rendering', () => {
  test('works for notifcation messages', async () => {
    const pullRequest = {
      ...pullRequestOpened.pull_request,
      labels: [],
      requested_reviewers: [{ login: 'user1' }],
      requested_teams: [{ name: 'Test-team', slug: 'test-team' }],
    };
    const prMessage = new PullRequest({
      pullRequest,
      repository: pullRequestOpened.repository,
      eventType: 'pull_request.opened',
      sender: pullRequestOpened.sender,
      reviews: [
        { state: 'CHANGES_REQUESTED', user: { login: 'user2' } },
        { state: 'APPROVED', user: { login: 'user2' } },
      ],
    });
    const rendered = prMessage.getRenderedMessage();
    expect(rendered).toMatchSnapshot();
  });

  test('works for notifcation messages with statuses', async () => {
    const pullRequest = {
      ...pullRequestOpened.pull_request,
      labels: [],
    };
    const prMessage = new PullRequest({
      pullRequest,
      repository: pullRequestOpened.repository,
      eventType: 'pull_request.opened',
      unfurl: false,
      statuses: combinedStatus.statuses,
      sender: pullRequestOpened.sender,
    });
    const rendered = prMessage.getRenderedMessage();
    expect(rendered).toMatchSnapshot();
  });

  test('works for notifcation messages with some statuses passing', async () => {
    const pullRequest = {
      ...pullRequestOpened.pull_request,
      labels: [],
    };
    const prMessage = new PullRequest({
      pullRequest,
      repository: pullRequestOpened.repository,
      eventType: 'pull_request.opened',
      unfurl: false,
      statuses: combinedStatusSomePassing.statuses,
      sender: pullRequestOpened.sender,
    });
    const rendered = prMessage.getRenderedMessage();
    expect(rendered).toMatchSnapshot();
  });

  test('works for notifcation messages with all statuses passing', async () => {
    const pullRequest = {
      ...pullRequestOpened.pull_request,
      labels: [],
    };
    const prMessage = new PullRequest({
      pullRequest,
      repository: pullRequestOpened.repository,
      eventType: 'pull_request.opened',
      unfurl: false,
      statuses: combinedStatusAllPassing.statuses,
      sender: pullRequestOpened.sender,
    });
    const rendered = prMessage.getRenderedMessage();
    expect(rendered).toMatchSnapshot();
  });

  test('works for notifcation messages with one status failing', async () => {
    const pullRequest = {
      ...pullRequestOpened.pull_request,
      labels: [],
    };
    const prMessage = new PullRequest({
      pullRequest,
      repository: pullRequestOpened.repository,
      eventType: 'pull_request.opened',
      unfurl: false,
      statuses: combinedStatusOneFailing.statuses,
      sender: pullRequestOpened.sender,
    });
    const rendered = prMessage.getRenderedMessage();
    expect(rendered).toMatchSnapshot();
  });

  test('works for minor notifications', async () => {
    const pullRequest = {
      ...pullRequestClosed.pull_request,
      labels: [],
    };
    const prMessage = new PullRequest({
      pullRequest,
      repository: pullRequestClosed.repository,
      eventType: 'pull_request.closed',
      sender: pullRequestClosed.sender,
    });
    const rendered = prMessage.getRenderedMessage();
    expect(rendered).toMatchSnapshot();
  });

  test('works for full unfurls', async () => {
    const pullRequest = {
      ...pullRequestOpened.pull_request,
      labels: [],
    };
    const prMessage = new PullRequest({
      pullRequest,
      repository: pullRequestOpened.repository,
      unfurlType: 'full',
    });
    const rendered = prMessage.getRenderedMessage();
    expect(rendered).toMatchSnapshot();
  });

  test('works for condensed unfurls', async () => {
    const pullRequest = {
      ...pullRequestOpened.pull_request,
      labels: [],
    };
    const prMessage = new PullRequest({
      pullRequest,
      repository: pullRequestOpened.repository,
      unfurlType: 'condensed',
    });
    const rendered = prMessage.getRenderedMessage();
    expect(rendered).toMatchSnapshot();
  });
});
