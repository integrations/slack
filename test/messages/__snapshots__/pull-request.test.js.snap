// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pull request rendering works for condensed unfurls 1`] = `
Object {
  "color": "#36a64f",
  "fallback": "[github-slack/app] #31 Add badges to readme",
  "footer": "<https://github.com/github-slack/app|github-slack/app>",
  "footer_icon": "https://assets-cdn.github.com/favicon.ico",
  "mrkdwn_in": Array [
    "text",
  ],
  "title": "#31 Add badges to readme",
  "title_link": "https://github.com/github-slack/app/pull/31",
  "ts": 1506092298,
}
`;

exports[`Pull request rendering works for full unfurls 1`] = `
Object {
  "author_icon": "https://avatars1.githubusercontent.com/u/7718702?v=4",
  "author_link": "https://github.com/wilhelmklopp",
  "author_name": "wilhelmklopp",
  "color": "#36a64f",
  "fallback": "[github-slack/app] #31 Add badges to readme",
  "fields": Array [
    Object {
      "short": true,
      "title": "Milestone",
      "value": "<https://github.com/github-slack/app/milestone/1|Staff Ship>",
    },
  ],
  "footer": "<https://github.com/github-slack/app|github-slack/app>",
  "footer_icon": "https://assets-cdn.github.com/favicon.ico",
  "mrkdwn_in": Array [
    "text",
  ],
  "text": "![image](https://user-images.githubusercontent.com/7718702/30750617-39e2fb52-9fb7-11e7-8da1-d6542317bee2.png)
",
  "title": "#31 Add badges to readme",
  "title_link": "https://github.com/github-slack/app/pull/31",
  "ts": 1506092298,
}
`;

exports[`Pull request rendering works for minor notifications 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#6f42c1",
      "fallback": "[github-slack/app] #36 Rearrange logic",
      "footer": "<https://github.com/github-slack/app|github-slack/app>",
      "footer_icon": "https://assets-cdn.github.com/favicon.ico",
      "mrkdwn_in": Array [
        "text",
      ],
      "pretext": "Pull request merged by bkeepers",
      "title": "#36 Rearrange logic",
      "title_link": "https://github.com/github-slack/app/pull/36",
      "ts": 1506619320,
    },
  ],
}
`;

exports[`Pull request rendering works for notification messages 1`] = `
Object {
  "attachments": Array [
    Object {
      "author_icon": "https://avatars1.githubusercontent.com/u/7718702?v=4",
      "author_link": "https://github.com/wilhelmklopp",
      "author_name": "wilhelmklopp",
      "color": "#36a64f",
      "fallback": "[github-slack/app] #31 Add badges to readme",
      "fields": Array [
        Object {
          "short": true,
          "title": "Reviewers",
          "value": "user2, @user1, @github-slack/test-team",
        },
        Object {
          "short": true,
          "title": "Milestone",
          "value": "<https://github.com/github-slack/app/milestone/1|Staff Ship>",
        },
      ],
      "footer": "<https://github.com/github-slack/app|github-slack/app>",
      "footer_icon": "https://assets-cdn.github.com/favicon.ico",
      "mrkdwn_in": Array [
        "text",
      ],
      "pretext": "Pull request opened by wilhelmklopp",
      "text": "![image](https://user-images.githubusercontent.com/7718702/30750617-39e2fb52-9fb7-11e7-8da1-d6542317bee2.png)
",
      "title": "#31 Add badges to readme",
      "title_link": "https://github.com/github-slack/app/pull/31",
      "ts": 1506092298,
    },
  ],
}
`;

exports[`Pull request rendering works for notification messages with all statuses passing 1`] = `
Object {
  "attachments": Array [
    Object {
      "author_icon": "https://avatars1.githubusercontent.com/u/7718702?v=4",
      "author_link": "https://github.com/wilhelmklopp",
      "author_name": "wilhelmklopp",
      "color": "#36a64f",
      "fallback": "[github-slack/app] #31 Add badges to readme",
      "fields": Array [
        Object {
          "short": true,
          "title": "Milestone",
          "value": "<https://github.com/github-slack/app/milestone/1|Staff Ship>",
        },
      ],
      "footer": "<https://github.com/github-slack/app|github-slack/app>",
      "footer_icon": "https://assets-cdn.github.com/favicon.ico",
      "mrkdwn_in": Array [
        "text",
      ],
      "pretext": "Pull request opened by wilhelmklopp",
      "text": "![image](https://user-images.githubusercontent.com/7718702/30750617-39e2fb52-9fb7-11e7-8da1-d6542317bee2.png)
",
      "title": "#31 Add badges to readme",
      "title_link": "https://github.com/github-slack/app/pull/31",
      "ts": 1506092298,
    },
    Object {
      "color": "#28a745",
      "fallback": "All checks have passed",
      "footer": "4/4 successful checks",
      "text": ":white_check_mark: All checks have passed",
    },
  ],
}
`;

exports[`Pull request rendering works for notification messages with check runs 1`] = `
Object {
  "attachments": Array [
    Object {
      "author_icon": "https://avatars1.githubusercontent.com/u/7718702?v=4",
      "author_link": "https://github.com/wilhelmklopp",
      "author_name": "wilhelmklopp",
      "color": "#36a64f",
      "fallback": "[github-slack/app] #31 Add badges to readme",
      "fields": Array [
        Object {
          "short": true,
          "title": "Milestone",
          "value": "<https://github.com/github-slack/app/milestone/1|Staff Ship>",
        },
      ],
      "footer": "<https://github.com/github-slack/app|github-slack/app>",
      "footer_icon": "https://assets-cdn.github.com/favicon.ico",
      "mrkdwn_in": Array [
        "text",
      ],
      "pretext": "Pull request opened by wilhelmklopp",
      "text": "![image](https://user-images.githubusercontent.com/7718702/30750617-39e2fb52-9fb7-11e7-8da1-d6542317bee2.png)
",
      "title": "#31 Add badges to readme",
      "title_link": "https://github.com/github-slack/app/pull/31",
      "ts": 1506092298,
    },
    Object {
      "author_icon": "https://github.com/images/error/octocat_happy.gif",
      "author_link": "http://github.com/github/hello-world/runs/4",
      "author_name": "Super CI: Mighty Readme report",
      "color": "#dbab09",
      "fallback": "Super CI: Mighty Readme report",
      "mrkdwn_in": Array [
        "text",
      ],
    },
    Object {
      "color": "#dbab09",
      "fallback": "2 other checks have passed",
      "footer": "2/3 successful checks",
      "text": ":white_check_mark: 2 other checks have passed",
    },
  ],
}
`;

exports[`Pull request rendering works for notification messages with one status failing 1`] = `
Object {
  "attachments": Array [
    Object {
      "author_icon": "https://avatars1.githubusercontent.com/u/7718702?v=4",
      "author_link": "https://github.com/wilhelmklopp",
      "author_name": "wilhelmklopp",
      "color": "#36a64f",
      "fallback": "[github-slack/app] #31 Add badges to readme",
      "fields": Array [
        Object {
          "short": true,
          "title": "Milestone",
          "value": "<https://github.com/github-slack/app/milestone/1|Staff Ship>",
        },
      ],
      "footer": "<https://github.com/github-slack/app|github-slack/app>",
      "footer_icon": "https://assets-cdn.github.com/favicon.ico",
      "mrkdwn_in": Array [
        "text",
      ],
      "pretext": "Pull request opened by wilhelmklopp",
      "text": "![image](https://user-images.githubusercontent.com/7718702/30750617-39e2fb52-9fb7-11e7-8da1-d6542317bee2.png)
",
      "title": "#31 Add badges to readme",
      "title_link": "https://github.com/github-slack/app/pull/31",
      "ts": 1506092298,
    },
    Object {
      "author_icon": "https://avatars2.githubusercontent.com/oa/1508?v=4",
      "author_link": "https://travis-ci.org/github-slack/app/builds/278661171?utm_source=github_status&utm_medium=notification",
      "author_name": "continuous-integration/travis-ci/pr: The Travis CI build is pending",
      "color": "#cb2431",
      "fallback": "continuous-integration/travis-ci/pr: The Travis CI build is pending",
      "mrkdwn_in": Array [
        "text",
      ],
    },
  ],
}
`;

exports[`Pull request rendering works for notification messages with some statuses passing 1`] = `
Object {
  "attachments": Array [
    Object {
      "author_icon": "https://avatars1.githubusercontent.com/u/7718702?v=4",
      "author_link": "https://github.com/wilhelmklopp",
      "author_name": "wilhelmklopp",
      "color": "#36a64f",
      "fallback": "[github-slack/app] #31 Add badges to readme",
      "fields": Array [
        Object {
          "short": true,
          "title": "Milestone",
          "value": "<https://github.com/github-slack/app/milestone/1|Staff Ship>",
        },
      ],
      "footer": "<https://github.com/github-slack/app|github-slack/app>",
      "footer_icon": "https://assets-cdn.github.com/favicon.ico",
      "mrkdwn_in": Array [
        "text",
      ],
      "pretext": "Pull request opened by wilhelmklopp",
      "text": "![image](https://user-images.githubusercontent.com/7718702/30750617-39e2fb52-9fb7-11e7-8da1-d6542317bee2.png)
",
      "title": "#31 Add badges to readme",
      "title_link": "https://github.com/github-slack/app/pull/31",
      "ts": 1506092298,
    },
    Object {
      "author_icon": "https://avatars2.githubusercontent.com/in/254?v=4",
      "author_link": "https://codecov.io/gh/github-slack/app/compare/0143c3cd70052e00bb07dc09229b261b3afd807d...68c2ef08b77a144aeaa69b601ecbacb979e3a8fd",
      "author_name": "codecov/patch: Coverage not affected when comparing 0143c3c...68c2ef0",
      "color": "#cb2431",
      "fallback": "codecov/patch: Coverage not affected when comparing 0143c3c...68c2ef0",
      "mrkdwn_in": Array [
        "text",
      ],
    },
    Object {
      "author_icon": "https://avatars2.githubusercontent.com/in/254?v=4",
      "author_link": "https://codecov.io/gh/github-slack/app/compare/0143c3cd70052e00bb07dc09229b261b3afd807d...68c2ef08b77a144aeaa69b601ecbacb979e3a8fd",
      "author_name": "codecov/project: 98.37% remains the same compared to 0143c3c",
      "color": "#dbab09",
      "fallback": "codecov/project: 98.37% remains the same compared to 0143c3c",
      "mrkdwn_in": Array [
        "text",
      ],
    },
    Object {
      "color": "#dbab09",
      "fallback": "2 other checks have passed",
      "footer": "2/4 successful checks",
      "text": ":white_check_mark: 2 other checks have passed",
    },
  ],
}
`;

exports[`Pull request rendering works for notification messages with statuses 1`] = `
Object {
  "attachments": Array [
    Object {
      "author_icon": "https://avatars1.githubusercontent.com/u/7718702?v=4",
      "author_link": "https://github.com/wilhelmklopp",
      "author_name": "wilhelmklopp",
      "color": "#36a64f",
      "fallback": "[github-slack/app] #31 Add badges to readme",
      "fields": Array [
        Object {
          "short": true,
          "title": "Milestone",
          "value": "<https://github.com/github-slack/app/milestone/1|Staff Ship>",
        },
      ],
      "footer": "<https://github.com/github-slack/app|github-slack/app>",
      "footer_icon": "https://assets-cdn.github.com/favicon.ico",
      "mrkdwn_in": Array [
        "text",
      ],
      "pretext": "Pull request opened by wilhelmklopp",
      "text": "![image](https://user-images.githubusercontent.com/7718702/30750617-39e2fb52-9fb7-11e7-8da1-d6542317bee2.png)
",
      "title": "#31 Add badges to readme",
      "title_link": "https://github.com/github-slack/app/pull/31",
      "ts": 1506092298,
    },
    Object {
      "author_icon": "https://avatars2.githubusercontent.com/oa/1508?v=4",
      "author_link": "https://travis-ci.org/github-slack/app/builds/278661171?utm_source=github_status&utm_medium=notification",
      "author_name": "continuous-integration/travis-ci/pr: The Travis CI build is pending",
      "color": "#dbab09",
      "fallback": "continuous-integration/travis-ci/pr: The Travis CI build is pending",
      "mrkdwn_in": Array [
        "text",
      ],
    },
    Object {
      "author_icon": "https://avatars2.githubusercontent.com/in/254?v=4",
      "author_link": "https://codecov.io/gh/github-slack/app/compare/0143c3cd70052e00bb07dc09229b261b3afd807d...68c2ef08b77a144aeaa69b601ecbacb979e3a8fd",
      "author_name": "codecov/patch: Coverage not affected when comparing 0143c3c...68c2ef0",
      "color": "#cb2431",
      "fallback": "codecov/patch: Coverage not affected when comparing 0143c3c...68c2ef0",
      "mrkdwn_in": Array [
        "text",
      ],
    },
    Object {
      "author_icon": "https://avatars2.githubusercontent.com/in/254?v=4",
      "author_link": "https://codecov.io/gh/github-slack/app/compare/0143c3cd70052e00bb07dc09229b261b3afd807d...68c2ef08b77a144aeaa69b601ecbacb979e3a8fd",
      "author_name": "codecov/project: 98.37% remains the same compared to 0143c3c",
      "color": "#cb2431",
      "fallback": "codecov/project: 98.37% remains the same compared to 0143c3c",
      "mrkdwn_in": Array [
        "text",
      ],
    },
    Object {
      "color": "#dbab09",
      "fallback": "1 other check has passed",
      "footer": "1/4 successful checks",
      "text": ":white_check_mark: 1 other check has passed",
    },
  ],
}
`;

exports[`Pull request rendering works for ready_for_review messages 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#36a64f",
      "fallback": "[github-slack/app] #31 Add badges to readme",
      "footer": "<https://github.com/github-slack/app|github-slack/app>",
      "footer_icon": "https://assets-cdn.github.com/favicon.ico",
      "mrkdwn_in": Array [
        "text",
      ],
      "pretext": "Pull request ready for review by wilhelmklopp",
      "title": "#31 Add badges to readme",
      "title_link": "https://github.com/github-slack/app/pull/31",
      "ts": 1506092298,
    },
  ],
}
`;
